src/app/games/abalone/tests/AbaloneRules.spec.ts

```diff
@@ -0,0 +1,438 @@
+        const state: AbaloneGameState = new AbaloneGameState(board, 0);
+
+        // When moving 3 pieces against 1 but then you're own piece block
+        const move: AbaloneMove = AbaloneMove.fromSingleCoord(new Coord(1, 4), HexaDirection.DOWN_RIGHT);
+        const status: AbaloneLegalityStatus = rules.isLegal(move, state);
+
+        // Then the move should be forbidden
+        expect(status.legal.reason).toBe(AbaloneFailure.CANNOT_PUSH_YOUR_OWN_PIECES);
+    });
+    it('should make pushed piece get of the board', () => {
+        // Given an board where 3 can push 1 out of 2 aligned pieces out of the board
+        const board: number[][] = [
+            [N, N, N, N, _, _, _, _, _],
+            [N, N, N, _, _, _, _, _, _],
+            [N, N, _, _, _, _, _, _, _],
+            [N, _, _, _, _, _, _, _, _],
+            [X, X, O, O, O, _, _, _, _],
+            [_, _, _, _, _, _, _, _, N],
+            [_, _, _, _, _, _, _, N, N],
+            [O, O, O, O, O, O, N, N, N],
+            [_, O, O, O, O, N, N, N, N],
+        ];
+        const state: AbaloneGameState = new AbaloneGameState(board, 0);
+
+        // When pushing
+        const move: AbaloneMove = AbaloneMove.fromSingleCoord(new Coord(4, 4), HexaDirection.UP_LEFT);
+        const status: AbaloneLegalityStatus = rules.isLegal(move, state);
+        const resultingState: AbaloneGameState = rules.applyLegalMove(move, state, status);
+
+        // Then the piece should be throwed out of the board
+        const expectedBoard: number[][] = [
+            [N, N, N, N, _, _, _, _, _],
+            [N, N, N, _, _, _, _, _, _],
+            [N, N, _, _, _, _, _, _, _],
+            [N, _, _, _, _, _, _, _, _],
+            [X, O, O, O, _, _, _, _, _],
+            [_, _, _, _, _, _, _, _, N],
+            [_, _, _, _, _, _, _, N, N],
+            [O, O, O, O, O, O, N, N, N],
+            [_, O, O, O, O, N, N, N, N],
+        ];
+        const expectedState: AbaloneGameState = new AbaloneGameState(expectedBoard, 1);
+        expect(status.legal.isSuccess()).toBeTrue();
+        expect(resultingState).toEqual(expectedState);
+    });
+    it('should declare player zero winner when he push a sixth enemy piece out of the board', () => {
+        const winningBoard: number[][] = [
+            [N, N, N, N, X, X, X, X, X],
+            [N, N, N, _, _, _, _, _, _],
+            [N, N, _, _, X, X, X, _, _],
+            [N, _, _, _, _, _, _, _, _],
+            [O, O, O, _, _, _, _, _, _],
+            [_, _, _, _, _, _, _, _, N],
+            [_, _, _, _, _, _, _, N, N],
+            [O, O, O, O, O, O, N, N, N],
+            [O, O, O, O, O, N, N, N, N],
+        ];
+        const winningState: AbaloneGameState = new AbaloneGameState(winningBoard, 1);
+        const node: AbaloneNode = new MGPNode(null, null, winningState);
+        expectToBeVictoryFor(rules, node, Player.ZERO, minimaxes);
+    });
+    it('should declare player one when he push a fifth enemy piece out of the board', () => {
```

Comment:
"player one" + winner
Tiens c'est encodé où le nombre de pièces enlevées ? Ou bien ça recalcule en parcourant le plateau à chaque fois.
Sinon : condition de victoire non symmétrique ?! (6th for player 0, 5th for player one)

Reply:
On compte le nombre de pièces restantes, c'est pas stocké.
5/6: fixed.